### PR TITLE
Split out the stable part of smir into its own crate to prevent accidental usage of forever unstable things

### PR DESCRIPTION
**Pull request by oli-obk**
Some groundwork for being able to work on https://github.com/rust-lang/project-stable-mir/issues/27 at all

r? @spastorino

### CODE REVIEW

File: compiler/smir/README.md [hunk 1]
```diff
@@ -0,0 +1,112 @@
...
+
+## Stable MIR Design
+
+The stable-mir will follow a similar approach to proc-macro2. It’s
+implementation will eventually be broken down into two main crates:
+
+- `stable_mir`: Public crate, to be published on crates.io, which will contain
+the stable data structure as well as proxy APIs to make calls to the
+compiler.
+- `rustc_smir`: The compiler crate that will translate from internal MIR to
```
Comment (Nilstrieb) [inline compiler/smir/README.md:85]: It would be nice to continue the parallels to proc macros here. Is this crate like proc_macro?

Comment (oli-obk) [reply compiler/smir/README.md:85]: not sure yet. We may add a proc-macro style intermediate API at some point

for now this is just a "we don't break you that often, and if we do, we have a transition API so you have a few months of grace period"-crate

Comment (oli-obk) [reply compiler/smir/README.md:85]: We're trying to figure this out. As the initial version I will add derives to the 0.1 release that then do 1:1 conversions from the unstable rustc version to the 0.1 version. When we plan a breaking change, we create a 0.2 version that does 1:1 conversion with an unstable equivalent. The 0.1 crate will get its compat layer changed to instead refer to the `0.2` crate and convert from 0.2 to 0.1 (semver trick).

Very vague and handwavy I know.

The alternative is to go the proc-macro route and just have a bunch of `extern "C"` functions to communicate with the compiler in a stable way, but I haven't given that too much thought yet, as the above mentioned system would be enough for most folk currently using unstable MIR

File: Cargo.lock [hunk 2]
```diff
@@ -4358,7 +4359,7 @@ dependencies = [
  "rustc_session",
  "rustc_span",
  "rustc_target",
- "scoped-tls",
+ "smir",
```
Comment (celinval) [inline Cargo.lock:4362]: What about `stable_mir`?

File: compiler/smir/README.md [hunk 3]
```diff
@@ -0,0 +1,112 @@
+This crate is regularly synced with its mirror in the rustc repo at `compiler/rustc_smir`.
```
Comment (celinval) [inline compiler/smir/README.md:1]: I'm assuming we are only going to mirror `compiler/stable_mir` from now on... is that the case?

**Review by celinval — COMMENTED** [review]
Thank you for doing this!

File: Cargo.lock (hunk 2 above)
Comment (oli-obk) [reply Cargo.lock:4362]: renamed

File: compiler/smir/README.md (hunk 3 above)
Comment (oli-obk) [reply compiler/smir/README.md]: Deduplicated the readme

**Pull request merged**